src/searcher/glue.rs

```diff
@@ -0,0 +1,262 @@
+use tree_sitter::{Node, QueryCursor};
+
+use crate::{
+    lines::{self, LineStep},
+    plugin::get_loaded_filter,
+    query_context::QueryContext,
+    searcher::{core::Core, Config, Range, Searcher},
+    sink::Sink,
+    treesitter::{get_parser, node_to_match},
+};
+
+#[derive(Debug, Default)]
+struct AccumulatedExactMatches {
+    matches_with_offsets_relative_to_reference_beginning_of_line_offset: Vec<Range>,
+    reference_beginning_of_line_offset: Option<usize>,
+}
+
+impl AccumulatedExactMatches {
+    pub fn clear(&mut self) {
+        self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+            .clear();
+        self.reference_beginning_of_line_offset = None;
+    }
+
+    pub fn push(
+        &mut self,
+        match_with_absolute_offsets: Range,
+        current_beginning_of_line_offset: usize,
+    ) {
+        if self.reference_beginning_of_line_offset.is_none() {
+            self.reference_beginning_of_line_offset = Some(current_beginning_of_line_offset);
+        }
+        self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+            .push(Range::new(
+                match_with_absolute_offsets.start()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+                match_with_absolute_offsets.end()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+            ));
+    }
+}
+
+impl AsRef<[Range]> for AccumulatedExactMatches {
+    fn as_ref(&self) -> &[Range] {
+        &self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+    }
+}
+
+#[derive(Debug)]
+pub struct MultiLine<'s, S> {
```

Comment:
We're always in "multi-line" mode from `grep`'s perspective but I kept most of the existing names the same (eg here it's not very "meaningful" that this is called `MultiLine`, but it is helpful for recognizing its correspondence to the same type in the `grep` source code)